File: examples/javascript/coding_agent.js
/**
 * Coding Agent - Run code analysis and tests in a sandbox.
 *
 * Usage:
 *   E2B_DOMAIN=localhost:8080 E2B_API_KEY=test-key node coding_agent.js
 */

const { Sandbox } = require("@e2b/code-interpreter");

async function main() {
  const sandbox = await Sandbox.create({
    apiKey: process.env.E2B_API_KEY || "test-key",
    domain: process.env.E2B_DOMAIN || "localhost:8080",
  });

  try {
    // 1. Set up environment
    console.log("1. Setting up dev environment...");
    await sandbox.commands.run("pip install pytest");

    // 2. Create project structure
    console.log("2. Creating project...");
    await sandbox.commands.run("mkdir -p /tmp/project/src /tmp/project/tests");

    await sandbox.files.write(
      "/tmp/project/src/calculator.py",
      `
class Calculator:
    def add(self, a, b):
        return a + b

    def subtract(self, a, b):
        return a - b

    def multiply(self, a, b):
        return a * b

    def divide(self, a, b):
        if b == 0:
            raise ValueError("Cannot divide by zero")
        return a / b
`
    );

    await sandbox.files.write(
      "/tmp/project/tests/test_calculator.py",
      `
import sys
sys.path.insert(0, '/tmp/project/src')
from calculator import Calculator

def test_add():
    c = Calculator()
    assert c.add(2, 3) == 5

def test_subtract():
    c = Calculator()
    assert c.subtract(5, 3) == 2

def test_multiply():
    c = Calculator()
    assert c.multiply(3, 4) == 12

def test_divide():
    c = Calculator()
    assert c.divide(10, 2) == 5.0
`
    );

    // 3. Run tests
    console.log("3. Running tests...");
    const result = await sandbox.commands.run("python -m pytest tests/ -v", {
      cwd: "/tmp/project",
    });
    console.log(`   Output:\n${result.stdout}`);
    console.log(`   Exit code: ${result.exitCode}`);

    // 4. Collect results
    console.log("4. Collecting results...");
    const execution = await sandbox.runCode(`
import json
report = {
    "project": "calculator",
    "status": "passed" if ${result.exitCode} == 0 else "failed",
    "exit_code": ${result.exitCode},
}
with open("/tmp/project/report.json", "w") as f:
    json.dump(report, f, indent=2)
print("Report generated")
`);
    for (const log of execution.logs) {
      console.log(`   ${log}`);
    }

    const report = await sandbox.files.read("/tmp/project/report.json");
    console.log(`   Report: ${report}`);
  } finally {
    await sandbox.kill();
    console.log("\nSandbox killed.");
  }
}

main().catch(console.error);

File: examples/python/coding_agent.py
"""Coding Agent - Run AI coding agent in a sandbox.

Demonstrates the coding agent workflow:
1. Create sandbox with development tools
2. Clone a repository
3. Run code analysis/linting
4. Execute tests
5. Collect results

Usage:
    export E2B_DOMAIN=localhost:8080
    export E2B_API_KEY=your-api-key
    python coding_agent.py
"""

import os
from e2b import Sandbox


def main():
    sandbox = Sandbox.create(
        template="base",
        api_key=os.environ.get("E2B_API_KEY", "test-key"),
        domain=os.environ.get("E2B_DOMAIN", "localhost:8080"),
    )

    try:
        # 1. Set up development environment
        print("1. Setting up dev environment...")
        sandbox.commands.run("pip install flake8 pytest")
        sandbox.envs.set({"PYTHONDONTWRITEBYTECODE": "1"})

        # 2. Create a project
        print("2. Creating sample project...")
        sandbox.commands.run("mkdir -p /tmp/project/src /tmp/project/tests")

        # Write source code
        sandbox.files.write("/tmp/project/src/calculator.py", """
class Calculator:
    def add(self, a, b):
        return a + b

    def subtract(self, a, b):
        return a - b

    def multiply(self, a, b):
        return a * b

    def divide(self, a, b):
        if b == 0:
            raise ValueError("Cannot divide by zero")
        return a / b
""")

        # Write tests
        sandbox.files.write("/tmp/project/tests/test_calculator.py", """
import sys
sys.path.insert(0, '/tmp/project/src')
from calculator import Calculator

def test_add():
    c = Calculator()
    assert c.add(2, 3) == 5
    assert c.add(-1, 1) == 0

def test_subtract():
    c = Calculator()
    assert c.subtract(5, 3) == 2

def test_multiply():
    c = Calculator()
    assert c.multiply(3, 4) == 12

def test_divide():
    c = Calculator()
    assert c.divide(10, 2) == 5.0

def test_divide_by_zero():
    c = Calculator()
    try:
        c.divide(1, 0)
        assert False, "Should have raised ValueError"
    except ValueError:
        pass
""")

        # 3. Run linting
        print("3. Running linter...")
        result = sandbox.commands.run(
            "flake8 /tmp/project/src/ --max-line-length=120",
            cwd="/tmp/project",
        )
        print(f"   Lint result: {'clean' if not result.stdout.strip() else result.stdout.strip()}")

        # 4. Run tests
        print("4. Running tests...")
        result = sandbox.commands.run(
            "python -m pytest tests/ -v",
            cwd="/tmp/project",
        )
        print(f"   Test output:\n{result.stdout}")
        print(f"   Exit code: {result.exit_code}")

        # 5. Generate report
        print("5. Generating report...")
        sandbox.run_code("""
import json

report = {
    "project": "calculator",
    "files": ["src/calculator.py", "tests/test_calculator.py"],
    "status": "all_tests_passed",
}
with open("/tmp/project/report.json", "w") as f:
    json.dump(report, f, indent=2)
print("Report generated")
""")

        report = sandbox.files.read("/tmp/project/report.json")
        print(f"   Report: {report}")

    finally:
        sandbox.kill()
        print("\nSandbox killed.")


if __name__ == "__main__":
    main()
